util/expensivequery/expensivequery.go

```diff
@@ -73,12 +81,208 @@ func (eqh *Handle) Run() {
 				}
 			}
 			threshold = atomic.LoadUint64(&variable.ExpensiveQueryTimeThreshold)
+			eqh.oomKillerAlert()
 		case <-eqh.exitCh:
 			return
 		}
 	}
 }
 
+var (
+	oomRecordErr      error
+	serverMemoryQuota uint64
+	oomRecordStatus   status // 0 uninitialized, 1 instance memory usage, 2 system memory usage, 3 close the alert.
+	lastOOMtime       time.Time
+
+	tmpDir              string
+	lastLogFileName     []string
+	lastProfileFileName []string
+	// OOMRecordCount counts the number of trigger oom record
+	OOMRecordCount uint32
+)
+
+type status uint32
+
+const (
+	uninitialized status = iota
+	useInstanceMemory
+	useSystemMemory
+	closeAlert
+)
+
+// If Performance.ServerMemoryQuota is set, use instance memory usage and ServerMemoryQuota * 80% to check oom risk.
+// If Performance.ServerMemoryQuota is not set, use system memory usage and total memory * 80% to check oom risk.
+func (eqh *Handle) oomKillerAlert() {
+	if oomRecordErr != nil || oomRecordStatus == closeAlert {
+		return
+	}
+
+	if oomRecordStatus == uninitialized {
+		if alert := config.GetGlobalConfig().Performance.ServerMemoryAlert; alert == 0 || alert == 1 {
+			oomRecordStatus = 3
+			return
+		}
+		if serverMemoryQuota == 0 {
+			if config.GetGlobalConfig().Performance.ServerMemoryQuota != 0 {
+				serverMemoryQuota = config.GetGlobalConfig().Performance.ServerMemoryQuota
+				oomRecordStatus = useInstanceMemory
+			} else {
+				serverMemoryQuota, oomRecordErr = memory.MemTotal()
+				if oomRecordErr != nil {
+					logutil.BgLogger().Warn("Get system total memory fail.", zap.Error(oomRecordErr))
+					return
+				}
+				oomRecordStatus = useSystemMemory
+			}
+		}
+		lastOOMtime = time.Time{}
+	}
+
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	if oomRecordStatus == 1 {
+		runtime.ReadMemStats(instanceStats)
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, oomRecordErr = memory.MemUsed()
+		if oomRecordErr != nil {
+			logutil.BgLogger().Warn("Get system usage memory fail.", zap.Error(oomRecordErr))
+			return
+		}
+	}
+
+	if float64(memoryUsage) > float64(serverMemoryQuota)*config.GetGlobalConfig().Performance.ServerMemoryAlert ||
+		(oomRecordStatus == useInstanceMemory && instanceStats.NextGC > serverMemoryQuota) {
+		// At least ten seconds between two recordings that memory usage is less than threshold (default 80% system memory).
+		// If the memory is still exceeded, only records once.
+		if time.Since(lastOOMtime) > 10*time.Second {
+			eqh.oomRecord(memoryUsage, serverMemoryQuota, oomRecordStatus == useSystemMemory)
+		}
+		lastOOMtime = time.Now()
+	}
+}
+
+func (eqh *Handle) oomRecord(memUsage uint64, serverMemoryQuota uint64, useSystemMemory bool) {
+	var err error
+	if tmpDir == "" {
+		tmpDir, err = ioutil.TempDir("", "TiDBOOM")
+		if err != nil {
+			return
+		}
+	}
+	tryRemove := func(filename []string) {
+		// Keep the last 5 files
+		if len(filename) < 5 {
```

Comment:
Can we make the number of keeping files manually?

Reply:
I think that adding another option to control the number of keeping files manually is unnecessary...